internal/pkg/composable/providers/kubernetes/hints.go

```diff
@@ -0,0 +1,234 @@
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package kubernetes
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/elastic/elastic-agent-autodiscover/utils"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+const (
+	integration = "package"
+	datastreams = "data_streams"
+
+	host        = "host"
+	period      = "period"
+	timeout     = "timeout"
+	metricspath = "metrics_path"
+	username    = "username"
+	password    = "password"
+)
+
+type hintsBuilder struct {
+	Key string
+
+	logger *logp.Logger
+}
+
+func (m *hintsBuilder) getIntegration(hints mapstr.M) string {
+	return utils.GetHintString(hints, m.Key, integration)
+}
+
+func (m *hintsBuilder) getDataStreams(hints mapstr.M) []string {
+	ds := utils.GetHintAsList(hints, m.Key, datastreams)
+	return ds
+}
+
+func (m *hintsBuilder) getHost(hints mapstr.M) string {
+	return utils.GetHintString(hints, m.Key, host)
+}
+
+func (m *hintsBuilder) getStreamHost(hints mapstr.M, streamName string) string {
+	key := fmt.Sprintf("%v.%v", streamName, host)
+	return utils.GetHintString(hints, m.Key, key)
+}
+
+func (m *hintsBuilder) getPeriod(hints mapstr.M) string {
+	return utils.GetHintString(hints, m.Key, period)
+}
+
+func (m *hintsBuilder) getStreamPeriod(hints mapstr.M, streamName string) string {
+	key := fmt.Sprintf("%v.%v", streamName, period)
+	return utils.GetHintString(hints, m.Key, key)
+}
+
+func (m *hintsBuilder) getTimeout(hints mapstr.M) string {
+	return utils.GetHintString(hints, m.Key, timeout)
+}
+
+func (m *hintsBuilder) getStreamTimeout(hints mapstr.M, streamName string) string {
+	key := fmt.Sprintf("%v.%v", streamName, timeout)
+	return utils.GetHintString(hints, m.Key, key)
+}
+
+func (m *hintsBuilder) getMetricspath(hints mapstr.M) string {
+	return utils.GetHintString(hints, m.Key, metricspath)
+}
+
+func (m *hintsBuilder) getStreamMetricspath(hints mapstr.M, streamName string) string {
+	key := fmt.Sprintf("%v.%v", streamName, metricspath)
+	return utils.GetHintString(hints, m.Key, key)
+}
+
+func (m *hintsBuilder) getUsername(hints mapstr.M) string {
+	return utils.GetHintString(hints, m.Key, username)
+}
+
+func (m *hintsBuilder) getStreamUsername(hints mapstr.M, streamName string) string {
+	key := fmt.Sprintf("%v.%v", streamName, username)
+	return utils.GetHintString(hints, m.Key, key)
+}
+
+func (m *hintsBuilder) getPassword(hints mapstr.M) string {
+	return utils.GetHintString(hints, m.Key, password)
+}
+
+func (m *hintsBuilder) getStreamPassword(hints mapstr.M, streamName string) string {
+	key := fmt.Sprintf("%v.%v", streamName, password)
+	return utils.GetHintString(hints, m.Key, key)
+}
+
+func (m *hintsBuilder) getFromMeta(value string, kubeMeta mapstr.M) string {
+	if value == "" {
+		return ""
+	}
+	r := regexp.MustCompile(`\${([^{}]+)}`)
```

Comment:
Why is this used?

Reply:
Within the hint's string you might have sth like `${kubernetes.pod.ip}` which you want to be retrieved dynamically from the Pod's meta. So here we have a regexp to extract the actual value of the variable we target out of a `"${kubernetes.*}"` hint.

To be more specific imagine that you have a hint like `'${kubernetes.pod.ip}:6379'`, then you want to replace the `${kubernetes.pod.ip}` part with value from the Pod's metadata so as to have sth like `10.28.90.345:6379`.

Reply:
added this as comment within the code too;) : [8b50e76](https://github.com/elastic/elastic-agent/pull/698/commits/8b50e766ed354692602f37c26a448dcc44abc044)